File: server/notes_streaming_service.py
"""
Voice Notes Capture Service - FINAL PRODUCTION VERSION (Correct Turn Boundaries + Summaries)
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from core_utils import (
    BaseStreamServer,
    stream_logger,
    MODEL,
    SEND_SAMPLE_RATE,
    PROJECT_ID,
    LOCATION,
)

import asyncio
import json
import base64
import uuid
from datetime import datetime
from pathlib import Path
import time

# Set Google Cloud environment variables
if PROJECT_ID:
    os.environ["GOOGLE_CLOUD_PROJECT"] = PROJECT_ID
if LOCATION:
    os.environ["GOOGLE_CLOUD_LOCATION"] = LOCATION
    os.environ["GOOGLE_CLOUD_REGION"] = LOCATION

from google.adk.agents import Agent, LiveRequestQueue
from google.adk.runners import Runner
from google.adk.agents.run_config import RunConfig, StreamingMode
from google.adk.sessions.in_memory_session_service import InMemorySessionService
from google.adk.errors.already_exists_error import AlreadyExistsError
from google.genai import types
from dotenv import load_dotenv
import websockets

load_dotenv()

# Notes storage directory
NOTES_DIR = "voice_notes_data"
Path(NOTES_DIR).mkdir(exist_ok=True)

# System instruction
NOTES_SYSTEM_INSTRUCTION = """You are a Voice Notes Assistant. Extract information from voice recordings.

When you receive audio, extract:
1. A title (3-8 words)
2. A summary (1-2 sentences)
3. Key points (1-5 items)

Then call save_voice_note immediately.

IMPORTANT: Always create a note, even if recording is brief or unclear.
If unclear: title="Quick Voice Note", summary="Brief recording", key_points=["Content was unclear"]

Always call save_voice_note right away after the audio turn ends.
"""


class NotesStreamingService(BaseStreamServer):
    """Voice notes service with properly managed turn boundaries"""

    def __init__(self, host="0.0.0.0", port=8081):
        super().__init__(host, port)
        self.session_service = InMemorySessionService()
        stream_logger.info("✓ Notes Streaming Service initialized on port 8081")

    async def manage_connection(self, websocket):
        connection_id = id(websocket)
        stream_logger.info(f"New notes connection: {connection_id}")
        await websocket.send(json.dumps({"type": "ready"}))
        
        try:
            await self.handle_notes_stream(websocket, connection_id)
        except websockets.exceptions.ConnectionClosed:
            stream_logger.info(f"Connection closed: {connection_id}")
        except Exception as e:
            stream_logger.error(f"Error: {e}")
            import traceback
            traceback.print_exc()
        finally:
            self.active_connections.pop(connection_id, None)

    async def handle_notes_stream(self, websocket, client_id):
        # Buffers
        self._input_texts = []
        self._is_recording = False
        self._audio_count = 0
        self._start_time = 0

        self.active_connections[client_id] = websocket

        user_id = f"notes_user_{client_id}"
        session_id = f"notes_session_{uuid.uuid4().hex[:16]}"

        # Create ADK session
        try:
            await self.session_service.create_session(
                app_name="notes_assistant",
                user_id=user_id,
                session_id=session_id,
            )
        except AlreadyExistsError:
            session_id = f"notes_session_{uuid.uuid4().hex[:16]}"
            await self.session_service.create_session(
                app_name="notes_assistant",
                user_id=user_id,
                session_id=session_id,
            )

        # TOOL DEFINITION
        async def save_voice_note_tool(title: str, summary: str, key_points: list[str]):
            stream_logger.info(f"🎯 TOOL CALLED: Saving note: {title}")

            note_id = f"note_{uuid.uuid4().hex[:12]}"
            timestamp = datetime.now().isoformat()

            note_data = {
                "id": note_id,
                "title": title,
                "summary": summary,
                "key_points": key_points,
                "timestamp": timestamp,
                "transcript": " ".join(self._input_texts),
            }

            filepath = os.path.join(NOTES_DIR, f"{note_id}.json")
            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(note_data, f, indent=2, ensure_ascii=False)

            try:
                await websocket.send(json.dumps({
                    "type": "note_saved",
                    "data": note_data
                }))
            except:
                pass

            return f"Saved {note_id}"

        # Create agent
        agent = Agent(
            name="voice_notes_agent",
            model=MODEL,
            instruction=NOTES_SYSTEM_INSTRUCTION,
            tools=[save_voice_note_tool],
        )

        runner = Runner(
            app_name="notes_assistant",
            agent=agent,
            session_service=self.session_service,
        )

        live_request_queue = LiveRequestQueue()
        audio_queue = asyncio.Queue()

        run_config = RunConfig(
            streaming_mode=StreamingMode.BIDI,
            input_audio_transcription=types.AudioTranscriptionConfig(),
        )

        # This will store the summarization prompt text
        self._pending_prompt = None
        live = None
        self._stop_stream = asyncio.Event()
        async with asyncio.TaskGroup() as tg:

            # RECEIVE MESSAGES FROM FRONTEND
            async def receive_client_messages():
                async for raw in websocket:
                    try:
                        data = json.loads(raw)

                        # ✔ START RECORDING → start a new TURN
                        if data.get("type") == "start_recording":
                            stream_logger.info("🎙️ START_RECORDING received")

                            self._is_recording = True
                            self._stop_stream.clear()
                            self._audio_count = 0
                            self._input_texts = []
                            self._start_time = time.time()

                            # OPEN TURN
                            live_request_queue.send_content(
                                types.Content(
                                    role="user",
                                    parts=[types.Part(text="Begin voice note now.")]
                                )
                            )
                            continue

                        # ✔ AUDIO CHUNK
                        if data.get("type") == "audio":
                            if self._is_recording is False:
                                continue

                            audio_bytes = base64.b64decode(data["data"])
                            await audio_queue.put(audio_bytes)
                            continue

                        # ✔ STOP_RECORDING → close turn
                        if data.get("type") == "stop_recording":
                            stream_logger.info("🛑 STOP_RECORDING received")

                            self._is_recording = False
                            self._stop_stream.set()

                            transcript = " ".join(self._input_texts).strip()
                            if not transcript:
                                transcript = "[No speech detected]"

                            self._pending_prompt = (
                                f"Recording finished.\n\n"
                                f"Transcript: {transcript}\n\n"
                                f"Extract title, summary, and 1–5 key points. "
                                f"Call save_voice_note immediately."
                            )
                            continue


                        # ✔ GET NOTES
                        if data.get("type") == "get_notes":
                            all_notes = self.get_all_notes()
                            await websocket.send(json.dumps({
                                "type": "notes_list",
                                "data": all_notes
                            }))
                            continue

                    except Exception as e:
                        stream_logger.error(f"Receive error: {e}")
                        import traceback
                        traceback.print_exc()

            # SEND AUDIO TO GEMINI
            async def send_audio():
                while True:
                    pcm = await audio_queue.get()

                    if not self._is_recording and audio_queue.empty():
                        audio_queue.task_done()
                        stream_logger.info("🧹 Audio sender drained and stopped")
                        break

                    if not self._is_recording:
                        audio_queue.task_done()
                        continue

                    if len(pcm) < 320:
                        audio_queue.task_done()
                        continue

                    try:
                        live_request_queue.send_realtime(
                            types.Blob(
                                data=pcm,
                                mime_type=f"audio/pcm;rate={SEND_SAMPLE_RATE}"
                            )
                        )
                    except Exception as e:
                        stream_logger.error(f"Realtime audio error: {e}")

                    audio_queue.task_done()



            # RECEIVE MODEL RESPONSES
            async def receive_service():
                nonlocal live
                live = runner.run_live(
                    user_id=user_id,
                    session_id=session_id,
                    live_request_queue=live_request_queue,
                    run_config=run_config,
                )

                async for event in live:
                    if self._stop_stream.is_set():
                        stream_logger.info("🧹 Stream stopped cleanly")
                        break


                    # ✔ TRANSCRIPTION CHUNKS
                    if hasattr(event, "input_transcription") and event.input_transcription:
                        text = event.input_transcription.text
                        if text:
                            self._input_texts.append(text)
                            await websocket.send(json.dumps({
                                "type": "text",
                                "data": text
                            }))

                    # ✔ TURN COMPLETE → Now send summarization prompt
                    if event.turn_complete:
                        stream_logger.info("✅ TURN COMPLETE")

                        if self._pending_prompt:
                            live_request_queue.send_content(
                                types.Content(
                                    role="user",
                                    parts=[types.Part(text=self._pending_prompt)]
                                )
                            )
                            self._pending_prompt = None

                        await websocket.send(json.dumps({"type": "recording_complete"}))

            tg.create_task(receive_client_messages())
            tg.create_task(send_audio())
            tg.create_task(receive_service())

    def get_all_notes(self):
        notes = []
        if not os.path.exists(NOTES_DIR):
            return notes

        for file in os.listdir(NOTES_DIR):
            if file.startswith("note_") and file.endswith(".json"):
                try:
                    with open(os.path.join(NOTES_DIR, file), "r", encoding="utf-8") as f:
                        notes.append(json.load(f))
                except:
                    continue

        notes.sort(key=lambda x: x.get("timestamp", ""), reverse=True)
        return notes


async def main():
    server = NotesStreamingService()
    await server.start_server()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass
    except Exception as e:
        stream_logger.error(f"Fatal error: {e}")

File: server-ts/src/notes-streaming-service.ts
/**
 * Voice Notes Capture Service
 *
 * WebSocket server for real-time voice notes capture using Gemini Live API.
 * Handles audio streaming, transcription, and automatic note extraction.
 */

import { WebSocketServer, WebSocket } from 'ws';
import { v4 as uuidv4 } from 'uuid';
import * as fs from 'fs';
import * as path from 'path';
import {
  config,
  logger,
  validateEnvVars,
  NOTES_SYSTEM_INSTRUCTION
} from './config.js';
import { createLiveClient, GeminiLiveClient, type FunctionDeclaration, type ToolCall } from './gemini-live-client.js';
import { getAzureBlobStorage } from './azure-blob-storage-singleton.js';

// Notes storage directory — relative to consumer's project root
const NOTES_DIR = path.join(process.cwd(), 'voice_notes_data');

let _notesDirReady = false;
async function ensureNotesDir(): Promise<void> {
  if (_notesDirReady) return;
  await fs.promises.mkdir(NOTES_DIR, { recursive: true });
  _notesDirReady = true;
}

interface NoteData {
  id: string;
  title: string;
  summary: string;
  key_points: string[];
  timestamp: string;
  transcript: string;
}

interface NotesConnection {
  ws: WebSocket;
  liveClient: GeminiLiveClient | null;
  sessionId: string;
  userId: string;
  inputTexts: string[];
  isRecording: boolean;
  audioCount: number;
  startTime: number;
  pendingPrompt: string | null;
}

const activeConnections = new Map<string, NotesConnection>();

// Tool definitions for save_voice_note
const NOTES_TOOLS: FunctionDeclaration[] = [
  {
    name: 'save_voice_note',
    description: 'Save a voice note with extracted title, summary, and key points.',
    parameters: {
      type: 'object',
      properties: {
        title: {
          type: 'string',
          description: 'A concise title for the note (3-8 words)',
        },
        summary: {
          type: 'string',
          description: 'A brief summary of the note content (1-2 sentences)',
        },
        key_points: {
          type: 'array',
          description: 'List of key points from the recording (1-5 items)',
        },
      },
      required: ['title', 'summary', 'key_points'],
    },
  },
];

/**
 * Handle save_voice_note tool call
 */
async function handleSaveVoiceNote(
  connection: NotesConnection,
  args: Record<string, unknown>
): Promise<string> {
  const title = typeof args.title === 'string' ? args.title : 'Untitled Note';
  const summary = typeof args.summary === 'string' ? args.summary : '';
  const keyPoints = Array.isArray(args.key_points) ? args.key_points.map(String) : [];

  logger.info(`Tool called: save_voice_note with title='${title}'`);

  await ensureNotesDir();

  const noteId = `note_${uuidv4().slice(0, 12)}`;
  const timestamp = new Date().toISOString();

  const noteData: NoteData = {
    id: noteId,
    title,
    summary,
    key_points: keyPoints,
    timestamp,
    transcript: connection.inputTexts.join(' '),
  };

  // Save to file
  const filepath = path.join(NOTES_DIR, `${noteId}.json`);
  await fs.promises.writeFile(filepath, JSON.stringify(noteData, null, 2), 'utf-8');

  // Upload to Azure Blob Storage (non-blocking, never fails the main flow)
  try {
    const azure = getAzureBlobStorage();
    await azure.waitForInit();
    const blobPath = await azure.uploadVoiceNote({ ...noteData }, noteId);
    if (blobPath) {
      logger.info(`Voice note uploaded to Azure: ${blobPath}`);
    } else {
      logger.warn(`Voice note Azure upload skipped (storage not configured)`);
    }
  } catch (error) {
    logger.error('Failed to upload voice note to Azure:', error);
  }

  // Send note_saved event to client
  try {
    connection.ws.send(JSON.stringify({
      type: 'note_saved',
      data: noteData,
    }));
    logger.info(`Sent note_saved event for: ${noteId}`);
  } catch (error) {
    logger.error('Error sending note_saved event:', error);
  }

  return `Saved note: ${noteId}`;
}

/**
 * Get all saved notes
 */
async function getAllNotes(): Promise<NoteData[]> {
  const notes: NoteData[] = [];

  try {
    await fs.promises.access(NOTES_DIR);
  } catch {
    return notes;
  }

  const files = await fs.promises.readdir(NOTES_DIR);
  for (const file of files) {
    if (file.startsWith('note_') && file.endsWith('.json')) {
      try {
        const filepath = path.join(NOTES_DIR, file);
        const content = await fs.promises.readFile(filepath, 'utf-8');
        notes.push(JSON.parse(content));
      } catch {
        continue;
      }
    }
  }

  // Sort by timestamp (newest first)
  notes.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
  return notes;
}

/**
 * Handle incoming client messages
 */
async function handleClientMessage(
  connection: NotesConnection,
  message: string
): Promise<void> {
  try {
    const data = JSON.parse(message);

    switch (data.type) {
      case 'start_recording':
        logger.info('START_RECORDING received');
        connection.isRecording = true;
        connection.audioCount = 0;
        connection.inputTexts = [];
        connection.startTime = Date.now();
        connection.pendingPrompt = null;

        // Send initial content to open the turn
        if (connection.liveClient?.connected) {
          connection.liveClient.sendText('Begin voice note now.');
        }
        break;

      case 'audio':
        // Forward audio to Gemini Live API
        if (connection.isRecording && connection.liveClient?.connected) {
          const audioBytes = Buffer.from(data.data, 'base64');
          if (audioBytes.length >= 320) {
            connection.liveClient.sendAudio(audioBytes);
            connection.audioCount++;
          }
        }
        break;

      case 'stop_recording':
        logger.info('STOP_RECORDING received');
        connection.isRecording = false;

        const transcript = connection.inputTexts.join(' ').trim() || '[No speech detected]';

        // Set pending prompt for summarization
        connection.pendingPrompt = `Recording finished.

Transcript: ${transcript}

Extract title, summary, and 1-5 key points. Call save_voice_note immediately.`;
        break;

      case 'get_notes':
        // Return all notes
        const notes = await getAllNotes();
        connection.ws.send(JSON.stringify({
          type: 'notes_list',
          data: notes,
        }));
        break;

      case 'text':
        // Send text message to Gemini
        if (connection.liveClient?.connected) {
          const text = data.data as string;
          logger.info(`Received text from client: ${text}`);
          connection.liveClient.sendText(text);
        }
        break;

      default:
        logger.warn(`Unknown message type: ${data.type}`);
    }
  } catch (error) {
    logger.error('Error handling client message:', error);
  }
}

/**
 * Set up Gemini Live client event handlers for notes service
 */
function setupNotesLiveClientHandlers(connection: NotesConnection): void {
  const { liveClient, ws } = connection;
  if (!liveClient) return;

  // Handle ready event
  liveClient.on('ready', () => {
    logger.info('Gemini Live API ready for notes');
  });

  // Handle audio output (we can ignore this for notes)
  liveClient.on('audio', (event) => {
    // Notes service doesn't need to send audio back to client
    // But we could if we wanted voice feedback
  });

  // Handle text output
  liveClient.on('text', (event) => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'text',
        data: event.data,
      }));
    }
  });

  // Handle input transcription
  liveClient.on('input_transcription', (event) => {
    if (typeof event.data === 'string' && event.data.trim()) {
      connection.inputTexts.push(event.data);
      if (ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({
          type: 'text',
          data: event.data,
        }));
      }
    }
  });

  // Handle output transcription
  liveClient.on('output_transcription', (event) => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'text',
        data: event.data,
      }));
    }
  });

  // Handle turn complete
  liveClient.on('turn_complete', () => {
    logger.info('Turn complete');

    // If there's a pending prompt (after stop_recording), send it
    if (connection.pendingPrompt && liveClient.connected) {
      liveClient.sendText(connection.pendingPrompt);
      connection.pendingPrompt = null;
    }

    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'recording_complete',
      }));
    }
  });

  // Handle interruption
  liveClient.on('interrupted', () => {
    logger.info('Response interrupted');
  });

  // Handle tool calls
  liveClient.on('tool_call', async (event) => {
    const toolCalls = event.toolCalls || [];

    for (const toolCall of toolCalls) {
      logger.info(`Tool call: ${toolCall.name}`, toolCall.args);

      if (toolCall.name === 'save_voice_note') {
        const result = await handleSaveVoiceNote(connection, toolCall.args);
        liveClient.sendToolResponse(toolCall.id, { result });
      }
    }
  });

  // Handle errors
  liveClient.on('error', (event) => {
    logger.error('Gemini Live API error:', event.data);
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'error',
        data: event.data,
      }));
    }
  });
}

/**
 * Handle new WebSocket connection
 */
async function handleConnection(ws: WebSocket): Promise<void> {
  const connectionId = uuidv4();
  logger.info(`New notes connection established: ${connectionId}`);

  // Create connection object
  const connection: NotesConnection = {
    ws,
    liveClient: null,
    sessionId: `notes_session_${uuidv4().slice(0, 16)}`,
    userId: `notes_user_${connectionId}`,
    inputTexts: [],
    isRecording: false,
    audioCount: 0,
    startTime: 0,
    pendingPrompt: null,
  };

  activeConnections.set(connectionId, connection);

  // Send ready message
  ws.send(JSON.stringify({ type: 'ready' }));

  try {
    // Create and connect Gemini Live client
    connection.liveClient = createLiveClient(NOTES_SYSTEM_INSTRUCTION, NOTES_TOOLS);
    setupNotesLiveClientHandlers(connection);
    await connection.liveClient.connect();

    // Handle incoming messages
    ws.on('message', (message: Buffer) => {
      handleClientMessage(connection, message.toString());
    });

    // Handle connection close
    ws.on('close', () => {
      logger.info(`Notes connection closed: ${connectionId}`);
      if (connection.liveClient) {
        connection.liveClient.disconnect();
      }
      activeConnections.delete(connectionId);
    });

    // Handle errors
    ws.on('error', (error) => {
      logger.error(`WebSocket error for notes connection ${connectionId}:`, error);
    });
  } catch (error) {
    logger.error(`Failed to initialize notes connection ${connectionId}:`, error);
    ws.send(JSON.stringify({
      type: 'error',
      data: 'Failed to connect to AI service',
    }));
    ws.close();
    activeConnections.delete(connectionId);
  }
}

/**
 * Start the notes streaming service
 */
async function main(portOverride?: number): Promise<void> {
  try {
    // Validate environment
    validateEnvVars();
    logger.info('Environment validated');

    const port = portOverride || config.NOTES_SERVICE_PORT;

    // Create WebSocket server
    const wss = new WebSocketServer({
      port: port,
      maxPayload: 100 * 1024 * 1024, // 100MB max payload
    });

    logger.info(`Voice Notes Streaming Service starting on port ${port}`);

    wss.on('connection', handleConnection);

    wss.on('error', (error) => {
      logger.error('WebSocket server error:', error);
    });

    logger.info(`Server running on ws://0.0.0.0:${port}`);
  } catch (error) {
    logger.error('Failed to start notes server:', error);
    process.exit(1);
  }
}

/**
 * Start the voice notes streaming service
 * @param port - Optional port override (defaults to config.NOTES_SERVICE_PORT env var)
 * @returns Promise that resolves when server is started
 */
export async function startNotesService(port?: number): Promise<void> {
  await main(port);
}

// Only run if executed directly (not imported)
const isMainModule = import.meta.url === `file://${process.argv[1]}`;
if (isMainModule) {
  const { loadEnv } = await import('./config.js');
  loadEnv();
  startNotesService().catch((error) => {
    logger.error('Fatal error:', error);
    process.exit(1);
  });

  // Graceful shutdown
  const shutdown = () => {
    logger.info('Shutting down notes service...');
    for (const [id, conn] of activeConnections) {
      conn.liveClient?.disconnect();
      conn.ws.close();
      activeConnections.delete(id);
    }
    process.exit(0);
  };
  process.on('SIGTERM', shutdown);
  process.on('SIGINT', shutdown);
}
